diff_drive_controller/src/diff_drive_controller.cpp

```diff
@@ -0,0 +1,588 @@
+    odometry_message.pose.pose.orientation.y = orientation.y();
+    odometry_message.pose.pose.orientation.z = orientation.z();
+    odometry_message.pose.pose.orientation.w = orientation.w();
+    odometry_message.twist.twist.linear.x = odometry_.getLinear();
+    odometry_message.twist.twist.angular.z = odometry_.getAngular();
+    realtime_odometry_publisher_->unlockAndPublish();
+  }
+
+  if (odom_params_.enable_odom_tf && odometry_transform_publisher_->is_activated() &&
+    realtime_odometry_transform_publisher_->trylock())
+  {
+    auto & transform = realtime_odometry_transform_publisher_->msg_.transforms.front();
+    transform.header.stamp = current_time;
+    transform.transform.translation.x = odometry_.getX();
+    transform.transform.translation.y = odometry_.getY();
+    transform.transform.rotation.x = orientation.x();
+    transform.transform.rotation.y = orientation.y();
+    transform.transform.rotation.z = orientation.z();
+    transform.transform.rotation.w = orientation.w();
+    realtime_odometry_transform_publisher_->unlockAndPublish();
+  }
+
+  const auto dt = current_time - received_velocity_msg_ptr_->header.stamp;
+
+  // Brake if cmd_vel has timeout
+  if (dt > cmd_vel_timeout_) {
+    linear_command = 0.0;
+    angular_command = 0.0;
+  }
+
+  const auto update_dt = current_time - previous_update_timestamp_;
+  previous_update_timestamp_ = current_time;
+
+  auto & last_command = previous_commands_.back().twist;
+  auto & second_to_last_command = previous_commands_.front().twist;
+  limiter_linear_.limit(
+    linear_command, last_command.linear.x, second_to_last_command.linear.x,
+    update_dt.seconds());
+  limiter_angular_.limit(
+    angular_command, last_command.angular.z, second_to_last_command.angular.z, update_dt.seconds());
+
+  previous_commands_.pop();
+  previous_commands_.emplace(*received_velocity_msg_ptr_);
+
+  //    Publish limited velocity
+  if (publish_limited_velocity_ && limited_velocity_publisher_->is_activated() &&
+    realtime_limited_velocity_publisher_->trylock())
+  {
+    auto & limited_velocity_command = realtime_limited_velocity_publisher_->msg_;
+    limited_velocity_command.header.stamp = current_time;
+    limited_velocity_command.twist.linear.x = linear_command;
+    limited_velocity_command.twist.angular.z = angular_command;
+    realtime_limited_velocity_publisher_->unlockAndPublish();
+  }
+
+  if (received_velocity_msg_ptr_ == nullptr) {
+    RCLCPP_WARN(logger, "Velocity message received was a nullptr.");
+    return CONTROLLER_INTERFACE_RET_ERROR;
+  }
+
+  // Compute wheels velocities:
+  const double velocity_left = (linear_command - angular_command * wheel_separation / 2.0) /
+    left_wheel_radius;
+  const double velocity_right = (linear_command + angular_command * wheel_separation / 2.0) /
+    right_wheel_radius;
+
+  // Set wheels velocities:
+  for (size_t index = 0; index < wheels.wheels_per_side; ++index) {
```

Comment:
better use an iterator or a "for each"

Reply:
it's tricky to do that while keeping this loop so compact...